src/validation.cpp

```diff
@@ -5915,7 +5909,7 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
                 // means <5MB of memory imprecision.
                 if (coins_processed % 120000 == 0) {
                     if (m_interrupt) {
-                        return false;
+                        return util::Error{};
```

Comment:
In commit "rpc: Return precise loadtxoutset error messages" (fa91404c68b4d503e6568f49aaf74a5746887185)

This is ok, but IMO it would be better to change result type from `Result<void>` to `Result<InterruptResult>` and return `Interrupted{}` here. If a function exits early because the user requested a shutdown, it may or may not be an error, and bubbling up `InterruptResult` lets the top level caller decide what to do based on context.

For example, in this case because the call is being made by RPC it would probably be good for the RPC call to return an "Snapshot loading was aborted" error. But in the GUI case, if a user tries to load snapshot and then shuts down before it is loaded, it would probably be better to just shut down cleanly and maybe log something, than try to present an error message.

In either case, returning something like `return util::Error{Untranslated("Processing snapshot aborted")}` or `return Interrupted{}` would be better than returning an error with no message.

Reply:
> An alternative may be to ask `Really shutdown and discard the loadutxoset progress? [Cancel] [Shutdown]`.

Yes exactly, that is the point of returning Interrupted status, because there are many ways caller may want to handle an interruption, and it should be up to the caller to decide. The behavior shouldn't be hardcoded into the function that was interrupted.

> But the GUI design can be done later, so I'll just clarify the error message here for the RPC.

That would be better than current version, but I wonder why you don't want to return interrupted status to make intent of the code clear and avoid needing to revisit this and change it later?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30497#discussion_r1693076584

> Overall I think it is not clear (at least to me) that `Result<InterruptResult>` is the way to go. Given the uncertainty it seems better to hold back on it for now. But that may just be me.

Great, thanks for explaining. You may want to add feedback on #29700 which makes extensive use of InterruptResult.

> I think Result<InterruptResult> would be wrong

It sounds like maybe you would want to this to `Result<void, InterruptResult>` instead of `Result<InterruptResult, void>`. I think that would be reasonable and could have advantages. I didn't opt for that approach because `return Interrupted{}` is shorter and seems more appropriate than `return {Error{}, Interrupted{}}`, because I wouldn't consider a successful interruption to be an error. But maybe it makes sense to support a util::Result constructor for failures that are not necessarily errors and allow `return Interrupted{}` syntax for `Result<void, InterruptResult>` types. I could experiment with that.

Reply:
> Your suggestion doesn't allow my suggestion. They are completely separate, because one is about cancelling the shutdown, before it happens, and the other is about returning a shutdown happened in loadtxoutset (after it happened).

Forgot to respond to this part. We would still want to return Interrupted in this case. To implement your suggestion the GUI would make an asynchronous ActivateSnapshot call when the user loaded a snapshot file, and show the `[Cancel] [Shutdown]` dialog if there was an attempt to shut down while the call was progress. The question is just what should ActivateSnapshot return after the user clicked [Shutdown], and `Interupted{}` makes more than `util::Error{Untranslated("Aborting after an interrupt was requested")}` to avoid triggering an error log or dialog when no error happened and the software did exactly what the user requested.